# BUILT-IN METHODS

31. hash(object)
32. input([prompt])
33. id(object)
34. isinstance(object, classinfo)
35. int(x=0, base=10)
36. issubclass(object, classinfo)
37. iter(object[, sentinel])
38. list([iterable])
39. locals()
40. len(s)
41. max(iterable, *iterables[,key, default])  / max(arg1, arg2, *args[, key])
42. min(iterable, *iterables[,key, default])  /  min(arg1, arg2, *args[, key])
43. map(function, iterable, ...)
44. next(iterator, default)
45. memoryview(obj)
46. object()
47. oct(x)
48. ord(c)
49. open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)
50. pow(x, y[, z])
51. print(*objects, sep=' ', end='\n', file=sys.stdout, flush=False)
52. property(fget=None, fset=None, fdel=None, doc=None)
53. range(stop) / range(start, stop[, step])
54. repr(obj)
55. reversed(seq)
56. round(number[, ndigits])
57. set([iterable])
58. setattr(object, name, value)
59. slice(stop) \ slice(start, stop, step)
60. sorted(iterable[, key][, reverse])
61. str(object='') / str(object=b'', encoding='utf-8', errors='strict')
62. sum(iterable, start)
63. tuple(iterable)
64. vars(object)
65. zip(*iterables)
66. __import__(name, globals=None, locals=None, fromlist=(), level=0)
67. super()

### 31. hash(object)
The hash() method returns the hash value of an object if it has one.

- Hash values are just integers which are used to compare dictionary keys during a dictionary lookup quickly.

- Internally, hash() method calls __hash__() method of an object which are set by default for any object. We'll look at this later.

In [3]:
# hash for integer unchanged
print('Hash for 181 is:', hash(181))

# hash for decimal
print('Hash for 181.23 is:',hash(181.23))

# hash for string
print('Hash for Python is:', hash('Python'))

Hash for 181 is: 181
Hash for 181.23 is: 530343892119126197
Hash for Python is: -6094861145673445078


In [4]:
# tuple of vowels
vowels = ('a', 'e', 'i', 'o', 'u')

print('The hash is:', hash(vowels))

The hash is: -8976551253661872762


In [5]:
class Person:
    def __init__(self, age, name):
        self.age = age
        self.name = name

    def __eq__(self, other):
        return self.age == other.age and self.name == other.name

    def __hash__(self):
        print('The hash is:')
        return hash((self.age, self.name))

person = Person(23, 'Adam')
print(hash(person))

The hash is:
698714698898671550


> Note: You don't have to implement the __eq__() method for the hash as it is created by default for all objects.

### 32. input([prompt])
The input() method reads a line from input, converts into a string and returns it.

In [6]:
inputString = input()

print('The inputted string is:', inputString)

test
The inputted string is: test


In [7]:
# get input from user

inputString = input('Enter a string:')

print('The inputted string is:', inputString)

Enter a string:Niket
The inputted string is: Niket


### 33. id(object)
The id() function returns identity (unique integer) of an object.

In [9]:
class Foo:
    b = 5

dummyFoo = Foo()
print('id of dummyFoo =',id(dummyFoo))

id of dummyFoo = 1992509487312


In [10]:
print('id of 5 =',id(5))

a = 5
print('id of a =',id(a))

b = a
print('id of b =',id(b))

c = 5.0
print('id of c =',id(c))

id of 5 = 1828023488
id of a = 1828023488
id of b = 1828023488
id of c = 1992498412160


> Note : It's important to note that everything in Python is an object, even numbers and Classes.
Hence, integer 5 has a unique id. The id of the integer 5 remains constant during the lifetime. Similar is the case for float 5.5 and other objects.

In [11]:
print(type(id(1)))
print(dir(id))

<class 'int'>
['__call__', '__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__name__', '__ne__', '__new__', '__qualname__', '__reduce__', '__reduce_ex__', '__repr__', '__self__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__text_signature__']


### 34. isinstance(object, classinfo)
The isinstance() function checks if the object (first argument) is an instance or subclass of classinfo class (second argument).

In [12]:
class Foo:
  a = 5
  
fooInstance = Foo()

print(isinstance(fooInstance, Foo))
print(isinstance(fooInstance, (list, tuple)))
print(isinstance(fooInstance, (list, tuple, Foo)))

True
False
True


In [13]:
numbers = [1, 2, 3]

result = isinstance(numbers, list)
print(numbers,'instance of list?', result)

result = isinstance(numbers, dict)
print(numbers,'instance of dict?', result)

result = isinstance(numbers, (dict, list))
print(numbers,'instance of dict or list?', result)

number = 5

result = isinstance(number, list)
print(number,'instance of list?', result)

result = isinstance(number, int)
print(number,'instance of int?', result)

[1, 2, 3] instance of list? True
[1, 2, 3] instance of dict? False
[1, 2, 3] instance of dict or list? True
5 instance of list? False
5 instance of int? True


### 35. int(x=0, base=10)
The int() method returns an integer object from any number or string.

In [14]:
# integer
print("int(123) is:", int(123))

# float
print("int(123.23) is:", int(123.23))

# string
print("int('123') is:", int('123'))

int(123) is: 123
int(123.23) is: 123
int('123') is: 123


In [9]:
print(type(int()))
print(dir(int))

<class 'int'>
['__abs__', '__add__', '__and__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getnewargs__', '__gt__', '__hash__', '__index__', '__init__', '__init_subclass__', '__int__', '__invert__', '__le__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']


In [15]:
# binary 0b or 0B
print("For 1010, int is:", int('1010', 2))
print("For 0b1010, int is:", int('0b1010', 2))

# octal 0o or 0O
print("For 12, int is:", int('12', 8))
print("For 0o12, int is:", int('0o12', 8))

# hexadecimal
print("For A, int is:", int('A', 16))
print("For 0xA, int is:", int('0xA', 16))

For 1010, int is: 10
For 0b1010, int is: 10
For 12, int is: 10
For 0o12, int is: 10
For A, int is: 10
For 0xA, int is: 10


Internally, int() method calls an object's __int__() method.

So, even if an object isn't a number, you can convert the object into an integer object.

You can do this by overriding the __index__() and __int__() methods of the class to return a number.

These two methods should return the same value as older versions of Python uses __int__(), while newer uses __index__() method.

In [16]:
class Person:
    age = 23

    def __index__(self):
        return self.age
    
    def __int__(self):
        return self.age

person = Person()
print('int(person) is:', int(person))

int(person) is: 23


### 36. issubclass(object, classinfo)
The issubclass() function checks if the object argument (first argument) is a subclass of classinfo class (second argument).

In [1]:
class Polygon:
  def __init__(polygonType):
    print('Polygon is a ', polygonType)

class Triangle(Polygon):
  def __init__(self):
    Polygon.__init__('triangle')
    
print(issubclass(Triangle, Polygon))
print(issubclass(Triangle, list))
print(issubclass(Triangle, (list, Polygon)))
print(issubclass(Polygon, (list, Polygon)))

True
False
True
True


### 37. iter(object[, sentinel])
The iter() method returns an iterator for the given object.

In [33]:
print (type(iter([1,2])))
print (type(iter({1:'1',2:'2'})))
print (type(iter({1,2})))
print (type(iter((1,2))))

<class 'list_iterator'>
<class 'dict_keyiterator'>
<class 'set_iterator'>
<class 'tuple_iterator'>


In [2]:
# list of vowels
vowels = ['a', 'e', 'i', 'o', 'u']

vowelsIter = iter(vowels)

# prints 'a'
print(next(vowelsIter))

# prints 'e'
print(next(vowelsIter))

# prints 'i'
print(next(vowelsIter))

# prints 'o'
print(next(vowelsIter))

# prints 'u'
print(next(vowelsIter))

a
e
i
o
u


In [3]:
class PrintNumber:
    def __init__(self, max):
        self.max = max

    def __iter__(self):
        self.num = 0
        return self

    def __next__(self):
        if(self.num >= self.max):
            raise StopIteration
        self.num += 1
        return self.num

printNum = PrintNumber(3)

printNumIter = iter(printNum)

# prints '1'
print(next(printNumIter))

# prints '2'
print(next(printNumIter))

# prints '3'
print(next(printNumIter))

# raises StopIteration
print(next(printNumIter))

1
2
3


StopIteration: 

When you run the program, it will open mydata.txt in read mode.

Then, the iter(fp.readline, '') in the for loop calls readline (which reads each line in the text file) until the sentinel character '' (empty string) is reached.

### 38. list([iterable])
The list() constructor creates a list in Python.

In [12]:
# empty list
print(list())

# vowel string
vowelString = 'aeiou'
print(list(vowelString))

# vowel tuple
vowelTuple = ('a', 'e', 'i', 'o', 'u')
print(list(vowelTuple))

# vowel list
vowelList = ['a', 'e', 'i', 'o', 'u']
print(list(vowelList))

[]
['a', 'e', 'i', 'o', 'u']
['a', 'e', 'i', 'o', 'u']
['a', 'e', 'i', 'o', 'u']


In [15]:
print(dir(list))

['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']


In [13]:
# vowel set
vowelSet = {'a', 'e', 'i', 'o', 'u'}
print(list(vowelSet))

# vowel dictionary
vowelDictionary = {'a': 1, 'e': 2, 'i': 3, 'o':4, 'u':5}
print(list(vowelDictionary))

['o', 'e', 'a', 'u', 'i']
['a', 'e', 'i', 'o', 'u']


In [14]:
class PowTwo:
    def __init__(self, max):
        self.max = max
    
    def __iter__(self):
        self.num = 0
        return self
        
    def __next__(self):
        if(self.num >= self.max):
            raise StopIteration
        result = 2 ** self.num
        self.num += 1
        return result

powTwo = PowTwo(5)
powTwoIter = iter(powTwo)

print(list(powTwoIter))

[1, 2, 4, 8, 16]


### 39. locals()
The locals() method updates and returns a dictionary of the current local symbol table.

In [1]:
locals()

{'__name__': '__main__',
 '__doc__': 'Automatically created module for IPython interactive environment',
 '__package__': None,
 '__loader__': None,
 '__spec__': None,
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '_ih': ['', 'locals()'],
 '_oh': {},
 '_dh': ['C:\\Users\\shindnik\\MyProject\\python\\MyNotes'],
 'In': ['', 'locals()'],
 'Out': {},
 'get_ipython': <bound method InteractiveShell.get_ipython of <ipykernel.zmqshell.ZMQInteractiveShell object at 0x00000287A1E99A20>>,
 'exit': <IPython.core.autocall.ZMQExitAutocall at 0x287a1eeca58>,
 'quit': <IPython.core.autocall.ZMQExitAutocall at 0x287a1eeca58>,
 '_': '',
 '__': '',
 '___': '',
 '_i': '',
 '_ii': '',
 '_iii': '',
 '_i1': 'locals()'}

In [3]:
print(_dh)
print (type(_dh))

['C:\\Users\\shindnik\\MyProject\\python\\MyNotes']
<class 'list'>


In [4]:
def localsNotPresent():
    return locals()

def localsPresent():
    present = True
    return locals()

print('localsNotPresent:', localsNotPresent())
print('localsPresent:', localsPresent())

localsNotPresent: {}
localsPresent: {'present': True}


In [5]:
def localsPresent():
    present = True
    print(present)
    locals()['present'] = False;
    print(present)

localsPresent()

True
True


In [6]:
def localsPresent():
    present = True
    print(present)
    locals()['present'] = False;
    print(present)

localsPresent()

True
True


### 40. len(s)

The len() function returns the number of items (length) in an object.
    - s : a sequence (string, bytes, tuple, list, or range) or a collection (dictionary, set or frozen set)

In [7]:
testList = []
print(testList, 'length is', len(testList))

testList = [1, 2, 3]
print(testList, 'length is', len(testList))

testTuple = (1, 2, 3)
print(testTuple, 'length is', len(testTuple))

testRange = range(1, 10)
print('Length of', testRange, 'is', len(testRange))

[] length is 0
[1, 2, 3] length is 3
(1, 2, 3) length is 3
Length of range(1, 10) is 9


In [8]:
testString = ''
print('Length of', testString, 'is', len(testString))

testString = 'Python'
print('Length of', testString, 'is', len(testString))

# byte object
testByte = b'Python'
print('Length of', testByte, 'is', len(testByte))

testList = [1, 2, 3]

# converting to bytes object
testByte = bytes(testList)
print('Length of', testByte, 'is', len(testByte))

Length of  is 0
Length of Python is 6
Length of b'Python' is 6
Length of b'\x01\x02\x03' is 3


In [9]:
testSet = {1, 2, 3}
print(testSet, 'length is', len(testSet))

# Empty Set
testSet = set()
print(testSet, 'length is', len(testSet))

testDict = {1: 'one', 2: 'two'}
print(testDict, 'length is', len(testDict))

testDict = {}
print(testDict, 'length is', len(testDict))

testSet = {1, 2}
# frozenSet
frozenTestSet = frozenset(testSet)
print(frozenTestSet, 'length is', len(frozenTestSet))

{1, 2, 3} length is 3
set() length is 0
{1: 'one', 2: 'two'} length is 2
{} length is 0
frozenset({1, 2}) length is 2


In [10]:
testSet = {1, 2, 3}
print(testSet, 'length is', len(testSet))

# Empty Set
testSet = set()
print(testSet, 'length is', len(testSet))

testDict = {1: 'one', 2: 'two'}
print(testDict, 'length is', len(testDict))

testDict = {}
print(testDict, 'length is', len(testDict))

testSet = {1, 2}
# frozenSet
frozenTestSet = frozenset(testSet)
print(frozenTestSet, 'length is', len(frozenTestSet))

{1, 2, 3} length is 3
set() length is 0
{1: 'one', 2: 'two'} length is 2
{} length is 0
frozenset({1, 2}) length is 2


In [11]:
class Session:
    def __init__(self, number = 0):
      self.number = number
    
    def __len__(self):
      return self.number


# default length is 0
s1 = Session()
print(len(s1))

# giving custom length
s2 = Session(6)
print(len(s2))

0
6


### 41. max(iterable, *iterables[,key, default])
### max(arg1, arg2, *args[, key])

The max() method returns the largest element in an iterable or largest of two or more parameters.

max(iterable, *iterables[, key, default])
    - iterable - sequence (tuple, string), collection (set, dictionary) or an iterator object whose largest element is to be found
    - *iterables (Optional) - any number of iterables whose largest is to be found
    - key (Optional) - key function where the iterables are passed and comparison is performed based on its return value
    - default (Optional) - default value if the given iterable is empty

max(arg1, arg2, *args[, key])
 - arg1 - mandatory first object for comparison (could be number, string or other object)
 - arg2 - mandatory second object for comparison (could be number, string or other object)
 - *args​ (Optional) - other objects for comparison
 - key - key function where each argument is passed, and comparison is performed based on its return value

In [16]:
# using max(arg1, arg2, *args)
print('Maximum is:', max(1, 3, 2, 5, 4))

# using max(iterable)
num = [1, 3, 2, 8, 5, 10, 6]
print('Maximum is:', max(num))

Maximum is: 5
Maximum is: 10


In [17]:
def sumDigit(num):
    sum = 0
    while(num):
        sum += num % 10
        num = int(num / 10)
    return sum

# using max(arg1, arg2, *args, key)
print('Maximum is:', max(100, 321, 267, 59, 40, key=sumDigit))

# using max(iterable, key)
num = [15, 300, 2700, 821, 52, 10, 6]
print('Maximum is:', max(num, key=sumDigit))

Maximum is: 267
Maximum is: 821


In [23]:
num = [15, 300, 2700, 821]
num1 = [12, 2]
num2 = [34, 567, 78]
print(max(num, num1, num2))

[34, 567, 78]


In [18]:
num = [15, 300, 2700, 821]
num1 = [12, 2]
num2 = [34, 567, 78]

# using max(iterable, *iterables, key)
print('Maximum is:', max(num, num1, num2, key=len))

Maximum is: [15, 300, 2700, 821]


In [20]:
d = {1:'1',2:'2',3:'3'}
print(max(d))

3


In [25]:
d1 = {1:'1',2:'2',3:'3'}
d2 = {4:'4',5:'5',6:'7'}
print(max(d1, d2))

TypeError: '>' not supported between instances of 'dict' and 'dict'

### 42. min(iterable, *iterables[,key, default])
### min(arg1, arg2, *args[, key])
he min() method returns the smallest element in an iterable or smallest of two or more parameters.

In [26]:
# using min(arg1, arg2, *args)
print('Minimum is:', min(1, 3, 2, 5, 4))

# using min(iterable)
num = [3, 2, 8, 5, 10, 6]
print('Minimum is:', min(num))

Minimum is: 1
Minimum is: 2


In [27]:
def sumDigit(num):
    sum = 0
    while(num):
        sum += num % 10
        num = int(num / 10)
    return sum

# using min(arg1, arg2, *args, key)
print('Minimum is:', min(100, 321, 267, 59, 40, key=sumDigit))

# using min(iterable, key)
num = [15, 300, 2700, 821, 52, 10, 6]
print('Minimum is:', min(num, key=sumDigit))

Minimum is: 100
Minimum is: 10


In [1]:
num = [15, 300, 2700, 821]
num1 = [12, 2]
num2 = [34, 567, 78]

# using min(iterable, *iterables, key)
print('Minimum is:', min(num, num1, num2))

Minimum is: [12, 2]


In [28]:
num = [15, 300, 2700, 821]
num1 = [12, 2]
num2 = [34, 567, 78]

# using min(iterable, *iterables, key)
print('Minimum is:', min(num, num1, num2, key=len))

Minimum is: [12, 2]


> In this program, each iterable num, num1 and num2 is passed to the built-in method len().
Based on the result, i.e. length of each list, the list with minimum length is returned.

### 43. map(function, iterable, ...)
The map() function applies a given function to each item of an iterable (list, tuple etc.) and returns a list of the results.

In [40]:
def calculateSquare(n):
  return n*n

numbers = (1, 2, 3, 4)
result = map(calculateSquare, numbers)
print(result)

# converting map object to set
numbersSquare1 = set(result)
print(numbersSquare1)

numbersSquare2 = set(result)
print(numbersSquare2)



{16, 1, 4, 9}
set()


In [41]:
print(dir(result))

['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__lt__', '__ne__', '__new__', '__next__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__']


In [38]:
def calculateSquare(n):
  return n*n

numbers = (1, 2, 3, 4)
result = map(calculateSquare, numbers)
print(dir(result))

NameError: name 'calculateSquare' is not defined

In [16]:
def calculateSquare(n):
  return n*n

numbers = (1, 2, 3, 4)
result = map(calculateSquare, numbers)
print(result)

# converting map object to set
numbersSquare1 = set(result)
print(numbersSquare1)

result = map(calculateSquare, numbers)
numbersSquare2 = list(result)
print(numbersSquare2)

result = map(calculateSquare, numbers)
numbersSquare2 = tuple(result)
print(numbersSquare2)


{16, 1, 4, 9}
[1, 4, 9, 16]
(1, 4, 9, 16)


In [17]:
numbers = (1, 2, 3, 4)
result = map(lambda x: x*x, numbers)
print(result)

# converting map object to set
numbersSquare = set(result)
print(numbersSquare)

{16, 1, 4, 9}


In [14]:
print(list(map(lambda x: x*x, range(1,10))))

[1, 4, 9, 16, 25, 36, 49, 64, 81]


In [13]:

print(list( filter (lambda x : x != None,list(map(lambda x: x*x if x % 2 == 0 else None, range(1,100))))))

[4, 16, 36, 64, 100, 144, 196, 256, 324, 400, 484, 576, 676, 784, 900, 1024, 1156, 1296, 1444, 1600, 1764, 1936, 2116, 2304, 2500, 2704, 2916, 3136, 3364, 3600, 3844, 4096, 4356, 4624, 4900, 5184, 5476, 5776, 6084, 6400, 6724, 7056, 7396, 7744, 8100, 8464, 8836, 9216, 9604]


In [6]:
print (list( filter (lambda x : x%2 == 0,range(1,100))))

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98]


In [19]:

print( list(map(lambda x: x*x,list( filter(lambda x : x%2 == 0,range(1,100))))))

[4, 16, 36, 64, 100, 144, 196, 256, 324, 400, 484, 576, 676, 784, 900, 1024, 1156, 1296, 1444, 1600, 1764, 1936, 2116, 2304, 2500, 2704, 2916, 3136, 3364, 3600, 3844, 4096, 4356, 4624, 4900, 5184, 5476, 5776, 6084, 6400, 6724, 7056, 7396, 7744, 8100, 8464, 8836, 9216, 9604]


### 44. next(iterator, default)
The next() function returns the next item from the iterator.

-The next() returns the next item from the iterator.
If the iterator is exhausted, it returns default value (if provided).
If the default parameter is omitted and iterator is exhausted, it raises StopIteration.

In [20]:
random = [5, 9, 'cat']

# converting list to iterator
randomIterator = iter(random)
print(randomIterator)

# Output: 5
print(next(randomIterator))

# Output: 9
print(next(randomIterator))

# Output: 'cat'
print(next(randomIterator))

# This will raise Error
# iterator is exhausted
print(next(randomIterator))

5
9
cat


StopIteration: 

In [34]:
random = [5, 9]

# converting list to iterator
randomIterator = iter(random)

# Output: 5
print(next(randomIterator, '-1'))

# Output: 9
print(next(randomIterator, '-1'))

# randomIterator is exhausted
# Output: '-1'
print(next(randomIterator, '-1'))
print(next(randomIterator, '-1'))
print(next(randomIterator, '-1'))

5
9
-1
-1
-1


### 45. memoryview(obj)
The memoryview() method returns a memory view object of the given argument.

obj - object whose internal data is to be exposed.
obj must support buffer protocol (bytes, bytearray)

> **buffer protocol** provides a way to access the internal data of an object. This internal data is a memory array or a buffer.

>**Memory view** is a safe way to expose the buffer protocol in Python.
It allows you to access the internal buffers of an object by creating a memory view object.

We need to remember that whenever we perform some action on an object (call a function of an object, slice an array), we (or Python) need to create a copy of the object.

If we have a large data to work with (eg. binary data of an image), we would unnecessarily create copies of huge chunks of data, which serves almost no use.

Using buffer protocol, we can give another object access to use/modify the large data without copying it. This makes the program use less memory and increases the execution speed.

In [44]:
#random bytearray
randomByteArray = bytearray('ABC', 'utf-8')

mv = memoryview(randomByteArray)

# access memory view's zeroth index
print(mv[0])

# create byte from memory view
print(bytes(mv[0:2]))

# create list from memory view
print(list(mv[0:3]))

# create list from memory view
print(list(mv[:]))

65
b'AB'
[65, 66, 67]
[65, 66, 67]


In [45]:
#random bytearray
randomByteArray = bytearray('ABC', 'utf-8')
print('Before updation:', randomByteArray)

mv = memoryview(randomByteArray)

# update 1st index of mv to Z
mv[1] = 90
print('After updation:', randomByteArray)

Before updation: bytearray(b'ABC')
After updation: bytearray(b'AZC')


### 46. object()
This returns a featureless object which is a base for all classes.

In [46]:
test = object()

print(type(test))
print(dir(test))


<class 'object'>
['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__']


### 47. oct(x)
The oct() method takes an integer number and returns its octal representation. If the given number is an int, it must implement __index__() method to return an integer.

In [47]:
# decimal number
print('oct(10) is:', oct(10))

# binary number
print('oct(0b101) is:', oct(0b101))

# hexadecimal number
print('oct(0XA) is:', oct(0XA))

oct(10) is: 0o12
oct(0b101) is: 0o5
oct(0XA) is: 0o12


In [48]:
class Person:
    age = 23

    def __index__(self):
        return self.age

    def __int__(self):
        return self.age

person = Person()
print('The oct is:', oct(person))

The oct is: 0o27


### 48. ord(c)
The ord() method returns an integer representing Unicode code point for the given Unicode character.

>The ord() method is the inverse of chr().

In [49]:
# code point of integer
print(ord('5'))

# code point of alphabet 
print(ord('A'))

# code point of character
print(ord('$'))

53
65
36


### 49. open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)


The open() function opens the file (if possible) and returns a corresponding file object.

- file - path-like object (representing a file system path) giving the pathname
- mode (optional) - mode while opening a file. If not provided, it defaults to 'r' (open for reading in text mode). Available file modes are:
     - 'r'	Open a file for reading. (default)
     - 'w'	Open a file for writing. Creates a new file if it does not exist or truncates the file if it exists.
    - 'x'	Open a file for exclusive creation. If the file already exists, the operation fails.
    - 'a'	Open for appending at the end of the file without truncating it. Creates a new file if it does not exist.
    - 't'	Open in text mode. (default)
    - 'b'	Open in binary mode.
    - '+'	Open a file for updating (reading and writing)
- buffering (optional) - used for setting buffering policy
- encoding (optional) - name of the encoding to encode or decode the file
- errors (optional) - string specifying how to handle encoding/decoding errors
- newline​ (optional) - how newlines mode works (available values: None, ' ', '\n', 'r', and '\r\n'
- closefd (optional) - must be True (default) if given otherwise an exception will be raised
- opener (optional) - a custom opener; must return an open file descriptor

In [60]:
# opens test.text file of the current directory
f = open("copy.txt")

print(type(f))
print(dir(f))

f.close()

<class '_io.TextIOWrapper'>
['_CHUNK_SIZE', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__lt__', '__ne__', '__new__', '__next__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '_checkClosed', '_checkReadable', '_checkSeekable', '_checkWritable', '_finalizing', 'buffer', 'close', 'closed', 'detach', 'encoding', 'errors', 'fileno', 'flush', 'isatty', 'line_buffering', 'mode', 'name', 'newlines', 'read', 'readable', 'readline', 'readlines', 'seek', 'seekable', 'tell', 'truncate', 'writable', 'write', 'writelines']


### 50. pow(x, y[, z])
the pow() method returns x to the power of y. If the third argument (z) is given, it returns x to the power of y modulus z, i.e. pow(x, y) % z.

>The pow(x, y) is equivalent to: x ** y

In [54]:
# positive x, positive y (x**y)
print(pow(2, 2))

# negative x, positive y
print(pow(-2, 2))

# positive x, negative y (x**-y)
print(pow(2, -2))

# negative x, negative y
print(pow(-2, -2))

4
4
0.25
0.25


In [56]:
x = 7
y = 2
z = 5
print((x**y)%z)
print(pow(x, y, z))

4
4


### 51. print(*objects, sep=' ', end='\n', file=sys.stdout, flush=False)

The print() function prints the given object to the standard output device (screen) or to the text stream file.

- objects - object to the printed. * indicates that there may be more than one object
- sep - objects are separated by sep. Default value: ' '
- end - end is printed at last
- file - must be an object with write(string) method. If omitted it, sys.stdout will be used which prints objects on the screen.
- flush - If True, the stream is forcibly flushed. Default value: False

In [57]:
print("Python is fun.")

a = 5
# Two objects are passed
print("a =", a)

b = a
# Three objects are passed
print('a =', a, '= b')

Python is fun.
a = 5
a = 5 = b


In [58]:
a = 5
print("a =", a, sep='00000', end='\n\n\n')
print("a =", a, sep='0', end='')

a =000005


a =05

In [63]:
sourceFile = open('python.txt', 'w')
print('Pretty cool, huh!', file = sourceFile)

sourceFile.close()

print( open('python.txt').read())


Pretty cool, huh!



### 52. property(fget=None, fset=None, fdel=None, doc=None)
The property() method a returns a property attribute.

##### property() Parameters
- The property() method takes four optional parameters:

    - fget (Optional) - function for getting the attribute value
    - fset (Optional) - function for setting the attribute value
    - fdel (Optional) - function for deleting the attribute value
    - doc (Optional) - string that contains the documentation (docstring) for the attribute

In [65]:
class Person:
    def __init__(self, name):
        self._name = name

    def getName(self):
        print('Getting name')
        return self._name

    def setName(self, value):
        print('Setting name to ' + value)
        self._name = value

    def delName(self):
        print('Deleting name')
        del self._name

    # Set property to use getName, setName
    # and delName methods
    name = property(getName, setName, delName, 'Name property')


p = Person('Adam')
print(p.name)

p.name = 'John'


Getting name
Adam
Setting name to John


In [66]:
class Person:
    def __init__(self, name):
        self._name = name

    @property
    def name(self):
        print('Getting name')
        return self._name

    @name.setter
    def name(self, value):
        print('Setting name to ' + value)
        self._name = value

    @name.deleter
    def name(self):
        print('Deleting name')
        del self._name

p = Person('Adam')
print('The name is:', p.name)

p.name = 'John'

del p.name

Getting name
The name is: Adam
Setting name to John
Deleting name


### 53. range(stop)
### range(start, stop[, step])

The range() type returns an immutable sequence of numbers between the given start integer to the stop integer.

- (If no step) Step defaults to 1. Returns a sequence of numbers starting from start and ending at stop - 1.
- (if step is zero) Raises a ValueError exception
- (if step is non-zero) Checks if the value constraint is met and returns a sequence according to the formula
If it doesn't meet the value constraint, Empty sequence is returned.


In [67]:
# empty range
print(list(range(0)))

# using range(stop)
print(list(range(10)))

# using range(start, stop)
print(list(range(1, 10)))

[]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [68]:
start = 2
stop = 14
step = 2

print(list(range(start, stop, step)))

[2, 4, 6, 8, 10, 12]


In [69]:
start = 2
stop = -14
step = -2

print(list(range(start, stop, step)))

# value constraint not met
print(list(range(start, 14, step)))

[2, 0, -2, -4, -6, -8, -10, -12]
[]


### 54.repr(obj)
The repr() method returns a printable representation of the given object.


- The repr() method returns a printable representational string of the given object.
- It returns a string that would yield an object with the same value when passed to eval().
- eval(repr(obj)) = obj (approximation that holds true for most of built-in types)

In [70]:
var = 'foo'

repr(var)

"'foo'"

In [71]:
class Person:
    name = 'Adam'

    def __repr__(self):
        return repr(self.name)

repr(Person())

"'Adam'"

### 55. reversed(seq)

The reversed() method returns the reversed iterator of the given sequence.

In [73]:
print(type(reversed('niket')))

<class 'reversed'>


In [74]:
print(dir(reversed('niket')))

['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__length_hint__', '__lt__', '__ne__', '__new__', '__next__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__']


In [72]:
# for string
seqString = 'Python'
print(list(reversed(seqString)))

# for tuple
seqTuple = ('P', 'y', 't', 'h', 'o', 'n')
print(list(reversed(seqTuple)))

# for range
seqRange = range(5, 9)
print(list(reversed(seqRange)))

# for list
seqList = [1, 2, 4, 3, 5]
print(list(reversed(seqList)))

['n', 'o', 'h', 't', 'y', 'P']
['n', 'o', 'h', 't', 'y', 'P']
[8, 7, 6, 5]
[5, 3, 4, 2, 1]


In [75]:
class Vowels:
    vowels = ['a', 'e', 'i', 'o', 'u']

    def __reversed__(self):
        return reversed(self.vowels)

v = Vowels()
print(list(reversed(v)))

['u', 'o', 'i', 'e', 'a']


### 56. round(number[, ndigits])

The round() method returns the floating point number rounded off to the given ndigits digits after the decimal point. If no ndigits is provided, it rounds off the number to the nearest integer.

In [76]:
# for integers
print(round(10))

# for floating point
print(round(10.7))

# even choice
print(round(5.5))

10
11
6


In [77]:
print(round(2.665, 2))

# cannot be represented exactly as float
print(round(2.675, 2))

2.67
2.67


> Here, both numbers result to the same output 2.67, when 2.675 should have been rounded to 2.68.
This isn't a bug but it is because, most decimal numbers cannot be represented exactly as a float.

### 57. set([iterable])
The set() constructor constructs a Python set from the given iterable and returns it.


In [1]:
# empty set
print(set())

# from string
print(set('Python'))

# from tuple
print(set(('a', 'e', 'i', 'o', 'u')))

# from list
print(set(['a', 'e', 'i', 'o', 'u']))

# from range
print(set(range(5)))

set()
{'o', 'P', 'y', 'n', 'h', 't'}
{'o', 'e', 'i', 'u', 'a'}
{'o', 'e', 'i', 'u', 'a'}
{0, 1, 2, 3, 4}


### 58. setattr(object, name, value)

The setattr() method sets the value of given attribute of an object.
    - object - object whose attribute has to be set
    - name - string which contains the name of the attribute to be set
    - value - value of the attribute to be set

In [2]:
class Person:
    name = 'Adam'
    
p = Person()
print('Before modification:', p.name)

# setting name to 'John'
setattr(p, 'name', 'John')

print('After modification:', p.name)

Before modification: Adam
After modification: John


If the attribute is not found, setattr() creates a new attribute by the given name and value.

However, this is only possible if the object implements __dict__() method.

You can check if it does by using dir() method.

In [6]:
class Person:
    name = 'Adam'
    
p = Person()

# setting attribute name to None
setattr(p, 'name', None)
print('Name is:', p.name)

# setting an attribute not present
# in Person
setattr(p, 'age', 23)
print('Age is:', p.age)

print(dir (p))

p1 = Person()

print(dir (p1))


Name is: None
Age is: 23
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'age', 'name']
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'name']


### 59. slice(stop)
### slice(start, stop, step)
The slice() constructor creates a slice object representing the set of indices specified by range(start, stop, step).

In [7]:
# contains indices (0, 1, 2)
print(slice(3))

# contains indices (1, 3)
print(slice(1, 5, 2))

slice(None, 3, None)
slice(1, 5, 2)


In [8]:
pyString = 'Python'

# contains indices (0, 1, 2)
# i.e. P, y and t
sObject = slice(3)

print(pyString[sObject])

# contains indices (1, 3)
# i.e. y and h
sObject = slice(1, 5, 2)

print(pyString[sObject])

Pyt
yh


In [10]:
pyString = 'Python'

# contains indices (-1, -2, -3)
# i.e. n, o and h
sObject = slice(-1, -4, -1)

print(pyString[sObject])

noh


In [11]:
pyList = ['P', 'y', 't', 'h', 'o', 'n']
pyTuple = ('P', 'y', 't', 'h', 'o', 'n')

# contains indices (0, 1, 2)
# i.e. P, y and t
sObject = slice(3)

# slice a list
print(pyList[sObject])

# contains indices (1, 3)
# i.e. y and h
sObject = slice(1, 5, 2)

# slice a tuple
print(pyTuple[sObject])

['P', 'y', 't']
('y', 'h')


In [12]:
pyList = ['P', 'y', 't', 'h', 'o', 'n']
pyTuple = ('P', 'y', 't', 'h', 'o', 'n')

# contains indices (-1, -2, -3)
# i.e. n, o and h
sObject = slice(-1, -4, -1)

# slice a list
print(pyList[sObject])

# contains indices (-1, -3)
# i.e. n and h
sObject = slice(-1, -5, -2)

# slice a tuple
print(pyTuple[sObject])

['n', 'o', 'h']
('n', 'h')


In [13]:
pyString = 'Python'

# contains indices (0, 1, 2)
# i.e. P, y and t
print(pyString[0:3])

# contains indices (1, 3)
# i.e. y and h
print(pyString[1:5:2])

Pyt
yh


### 60. sorted(iterable[, key][, reverse])
The sorted() method sorts the elements of a given iterable in a specific order - Ascending or Descending.
    - iterable - sequence (string, tuple, list) or collection (set, dictionary, frozen set) or any iterator 
    - reverse (Optional) - If true, the sorted list is reversed (or sorted in Descending order)
    - key (Optional) - function that serves as a key for the sort comparison

In [14]:
# vowels list
pyList = ['e', 'a', 'u', 'o', 'i']
print(sorted(pyList))

# string 
pyString = 'Python'
print(sorted(pyString))

# vowels tuple
pyTuple = ('e', 'a', 'u', 'o', 'i')
print(sorted(pyTuple))

['a', 'e', 'i', 'o', 'u']
['P', 'h', 'n', 'o', 't', 'y']
['a', 'e', 'i', 'o', 'u']


In [15]:
# set
pySet = {'e', 'a', 'u', 'o', 'i'}
print(sorted(pySet, reverse=True))

# dictionary
pyDict = {'e': 1, 'a': 2, 'u': 3, 'o': 4, 'i': 5}
print(sorted(pyDict, reverse=True))

# frozen set
pyFSet = frozenset(('e', 'a', 'u', 'o', 'i'))
print(sorted(pyFSet, reverse=True))

['u', 'o', 'i', 'e', 'a']
['u', 'o', 'i', 'e', 'a']
['u', 'o', 'i', 'e', 'a']


In [16]:
# take second element for sort
def takeSecond(elem):
    return elem[1]

# random list
random = [(2, 2), (3, 4), (4, 1), (1, 3)]

# sort list with key
sortedList = sorted(random, key=takeSecond)

# print list
print('Sorted list:', sortedList)

Sorted list: [(4, 1), (2, 2), (1, 3), (3, 4)]


### 61. str(object='')
### str(object=b'', encoding='utf-8', errors='strict')

The str() method returns the "informal" or nicely printable representation of a given object.

    - object - object whose informal representation is to be returned
    - encoding - Defaults of UTF-8. Encoding of the given object
    - errors - response when decoding fails. There are six types of error response
    - strict - default response which raises a UnicodeDecodeError exception on failure
    - ignore - ignores the unencodable unicode from the result
    - replace - replaces the unencodable unicode to a question mark ?
    - xmlcharrefreplace - inserts XML character reference instead of unencodable unicode
    - backslashreplace - inserts a \uNNNN espace sequence instead of unencodable unicode
    - namereplace - inserts a \N{...} escape sequence instead of unencodable unicode

In [17]:
str('10')

'10'

In [18]:
# bytes
b = bytes('pythön', encoding='utf-8')

print(str(b, encoding='ascii', errors='ignore'))

pythn


### 62. sum(iterable, start)

The sum() function adds the items of an iterable and returns the sum.

In [19]:
numbers = [2.5, 3, 4, -5]

# start parameter is not provided
numbersSum = sum(numbers)
print(numbersSum)

# start = 10
numbersSum = sum(numbers, 10)
print(numbersSum)

4.5
14.5


### 63. tuple(iterable)
The tuple() built-in is used to create a tuple in Python.

In [20]:
t1 = tuple()
print('t1=', t1)

# creating a tuple from a list
t2 = tuple([1, 4, 6])
print('t2=', t2)

# creating a tuple from a string
t1 = tuple('Python')
print('t1=',t1)

# creating a tuple from a dictionary
t1 = tuple({1: 'one', 2: 'two'})
print('t1=',t1)

t1= ()
t2= (1, 4, 6)
t1= ('P', 'y', 't', 'h', 'o', 'n')
t1= (1, 2)


### 64. vars(object)

The vars() function returns the __dict__ attribute of the given object if the object has __dict__ attribute.

In [22]:
class Foo:
  def __init__(self, a = 5, b = 10):
    self.a = a
    self.b = b
  
InstanceOfFoo = Foo()
print(vars(InstanceOfFoo))

{'a': 5, 'b': 10}


In [25]:
vars(list)

mappingproxy({'__repr__': <slot wrapper '__repr__' of 'list' objects>,
              '__hash__': None,
              '__getattribute__': <slot wrapper '__getattribute__' of 'list' objects>,
              '__lt__': <slot wrapper '__lt__' of 'list' objects>,
              '__le__': <slot wrapper '__le__' of 'list' objects>,
              '__eq__': <slot wrapper '__eq__' of 'list' objects>,
              '__ne__': <slot wrapper '__ne__' of 'list' objects>,
              '__gt__': <slot wrapper '__gt__' of 'list' objects>,
              '__ge__': <slot wrapper '__ge__' of 'list' objects>,
              '__iter__': <slot wrapper '__iter__' of 'list' objects>,
              '__init__': <slot wrapper '__init__' of 'list' objects>,
              '__len__': <slot wrapper '__len__' of 'list' objects>,
              '__getitem__': <method '__getitem__' of 'list' objects>,
              '__setitem__': <slot wrapper '__setitem__' of 'list' objects>,
              '__delitem__': <slot wrapper '__del

### 65. zip(*iterables)

The zip() function take iterables (can be zero or more), makes iterator that aggregates elements based on the iterables passed, and returns an iterator of tuples.

In [29]:
print(type(zip()))

<class 'zip'>


In [30]:
print(dir(zip))

['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__lt__', '__ne__', '__new__', '__next__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__']


In [28]:
numberList = [1, 2, 3]
strList = ['one', 'two', 'three']

# No iterables are passed
result = zip()

# Converting itertor to list
resultList = list(result)
print(resultList)

# Two iterables are passed
result = zip(numberList, strList)

# Converting itertor to set
resultSet = set(result)
print(resultSet)

[]
{(3, 'three'), (2, 'two'), (1, 'one')}


In [31]:
numbersList = [1, 2, 3]
strList = ['one', 'two']
numbersTuple = ('ONE', 'TWO', 'THREE', 'FOUR')

result = zip(numbersList, numbersTuple)

# Converting to set
resultSet = set(result)
print(resultSet)

result = zip(numbersList, strList, numbersTuple)

# Converting to set
resultSet = set(result)

{(2, 'TWO'), (1, 'ONE'), (3, 'THREE')}


In [32]:
coordinate = ['x', 'y', 'z']
value = [3, 4, 5, 0, 9]

result = zip(coordinate, value)
resultList = list(result)
print(resultList)

c, v =  zip(*resultList)
print('c =', c)
print('v =', v)

[('x', 3), ('y', 4), ('z', 5)]
c = ('x', 'y', 'z')
v = (3, 4, 5)


66. __import__(name, globals=None, locals=None, fromlist=(), level=0)

The __import__() is an advanced function that is called by the import statement.
   
   - name - name of the module you want to import
   - globals and locals - determines how to interpert name
   - fromlist - objects or submodules that should be imported by name
   - level - specifies whether to use absolute or relative imports

In [33]:
mathematics = __import__('math', globals(), locals(), [], 0)

print(mathematics.fabs(-2.5))

2.5


### 67. super()
The super() builtin returns a proxy object that allows you to refer parent class by 'super'.

In Python, super() built-in has two major use cases:

        - Allows us to avoid using base class explicitly
        - Working with Multiple Inheritance

In [1]:
class Mammal(object):
  def __init__(self, mammalName):
    print(mammalName, 'is a warm-blooded animal.')
    
class Dog(Mammal):
  def __init__(self):
    print('Dog has four legs.')
    super().__init__('Dog')
    
d1 = Dog()

Dog has four legs.
Dog is a warm-blooded animal.


In [3]:
class Animal:
  def __init__(self, animalName):
    print(animalName, 'is an animal.');

class Mammal(Animal):
  def __init__(self, mammalName):
    print(mammalName, 'is a warm-blooded animal.')
    super().__init__(mammalName)
    
class NonWingedMammal(Mammal):
  def __init__(self, NonWingedMammalName):
    print(NonWingedMammalName, "can't fly.")
    super().__init__(NonWingedMammalName)

class NonMarineMammal(Mammal):
  def __init__(self, NonMarineMammalName):
    print(NonMarineMammalName, "can't swim.")
    super().__init__(NonMarineMammalName)

class Dog(NonMarineMammal, NonWingedMammal):
  def __init__(self):
    print('Dog has 4 legs.');
    super().__init__('Dog')
    
d = Dog()
print('')


bat = NonMarineMammal('Bat') 

Dog has 4 legs.
Dog can't swim.
Dog can't fly.
Dog is a warm-blooded animal.
Dog is an animal.

Bat can't swim.
Bat is a warm-blooded animal.
Bat is an animal.


#### ______________________________Method Resolution Order (MRO)__________________________________________________

In [4]:
Dog.__mro__

(__main__.Dog,
 __main__.NonMarineMammal,
 __main__.NonWingedMammal,
 __main__.Mammal,
 __main__.Animal,
 object)

Here is how MRO is calculated in Python:

- A method in the derived calls is always called before the method of the base class. 
- In our example, Dog class is called before NonMarineMammal or NoneWingedMammal. These two classes are called before Mammal which is called before Animal, and Animal class is called before object.
- If there are multiple parents like Dog(NonMarineMammal, NonWingedMammal), method of NonMarineMammal is invoked first because it appears first.

In [2]:
class Animal:
  def __init__(self, animalName):
    print(animalName, 'is an animal.');

class Mammal(Animal):
  def __init__(self, mammalName):
    print(mammalName, 'is a warm-blooded animal.')
    super().__init__(mammalName)
    
class NonMarineMammal(Mammal):
  def __init__(self, NonMarineMammalName):
    print(NonMarineMammalName, "can't swim.")
    super().__init__(NonMarineMammalName)

class NonWingedMammal(Mammal):
  def __init__(self, NonWingedMammalName):
    print(NonWingedMammalName, "can't fly.")
    super().__init__(NonWingedMammalName)


class Dog(NonMarineMammal, NonWingedMammal):
  def __init__(self):
    print('Dog has 4 legs.');
    super().__init__('Dog')
    
d = Dog()
print('')


bat = NonMarineMammal('Bat') 

Dog has 4 legs.
Dog can't swim.
Dog can't fly.
Dog is a warm-blooded animal.
Dog is an animal.

Bat can't swim.
Bat is a warm-blooded animal.
Bat is an animal.
